### PR TITLE
[CVP] Flip signedness icmp predicate in use level

### LINKED ISSUE #69928: [CVP] Missed opportunity to do slt to ult conversion
For the following [example](https://godbolt.org/z/1PePcKve7), CVP is not converting `slt` to `ult`:
```
define i64 @test_select(i64 %arg, i64 %arg1) {
entry:
  %cmp1 = icmp ult i64 %arg, 64424509440
  %and = and i64 %arg1, 4294967295
  %cmp2 = icmp slt i64 %arg, %and
  %select = select i1 %cmp1, i1 %cmp2, i1 false
  br i1 %select, label %good, label %bad

bad:
  ret i64 1

good:
  ret i64 0
}
```
[Proof](https://alive2.llvm.org/ce/z/cq7ydm) that this transformation is valid.
On the other hand, it is [able](https://godbolt.org/z/44eq4o5vv) to convert it if there are 2 BBs and no select instruction.

### PR DESCRIPTION
**Pull request by XChy**
Resolve #69928.
[Alive2 Proof](https://alive2.llvm.org/ce/z/zoqaqf).

### CODE REVIEW

File: llvm/test/Transforms/CorrelatedValuePropagation/icmp.ll [hunk 1]
```diff
@@ -1246,3 +1246,78 @@ declare <2 x i8> @llvm.umin.v2i8(<2 x i8>, <2 x i8>)
...
+; CHECK:       bad:
+; CHECK-NEXT:    ret i64 1
+; CHECK:       good:
+; CHECK-NEXT:    ret i64 0
+;
+entry:
+  %cmp1 = icmp ult i64 %arg, 1000
+  %cmp2 = icmp slt i64 %arg, 100
+  %select = select i1 %cmp1, i1 %cmp2, i1 false
+  br i1 %select, label %good, label %bad
```
Comment (nikic) [inline llvm/test/Transforms/CorrelatedValuePropagation/icmp.ll:1266]: The br is unnecessary.

**Review by nikic — COMMENTED** [review]
The change looks good to me, but can you please adjust the test coverage so the second condition is not implied (i.e. folds to true) by the first? In that case the transform isn't really meaningful.

As you mark https://github.com/llvm/llvm-project/issues/69928 as fixed, you should also add the test case from there.

File: llvm/test/Transforms/CorrelatedValuePropagation/icmp.ll (hunk 1 above)
Comment (nikic) [reply llvm/test/Transforms/CorrelatedValuePropagation/icmp.ll:1268]: You've removed it everywhere but here?

File: llvm/test/Transforms/CorrelatedValuePropagation/icmp.ll [hunk 2]
```diff
@@ -1246,3 +1246,118 @@ declare <2 x i8> @llvm.umin.v2i8(<2 x i8>, <2 x i8>)
...
+
+bad:
+  ret i64 1
+
+good:
+  ret i64 0
+}
+
+define i1 @test_select_flip(i64 noundef %arg) {
```
Comment (nikic) [inline llvm/test/Transforms/CorrelatedValuePropagation/icmp.ll:1277]: Drop the noundef arguments, unless they are necessary.

**Review by nikic — APPROVED** [review]
LGTM apart from test nits.

File: llvm/test/Transforms/CorrelatedValuePropagation/icmp.ll (hunk 1 above)
Comment (XChy) [reply llvm/test/Transforms/CorrelatedValuePropagation/icmp.ll:1268]: It is a copy from the original issue, so it contains br.

File: llvm/test/Transforms/CorrelatedValuePropagation/icmp.ll (hunk 2 above)
Comment (XChy) [reply llvm/test/Transforms/CorrelatedValuePropagation/icmp.ll:1277]: `getConstantRangeAtUse` requires noundef/no poison. I don't get why, so keep this flag.

Comment (nikic) [reply llvm/test/Transforms/CorrelatedValuePropagation/icmp.ll:1277]: Ah yes, you are right. Some at-use transforms may be unsafe with undef values.

**Pull request merged**